### PR TITLE
implemented login, signup endpoints with jwt

### LINKED ISSUE #13: Implement JWT authentication
Reference:
https://tutorialedge.net/golang/authenticating-golang-rest-api-with-jwts/

### LINKED ISSUE #18: Sort the imports across the codebase
Use this format:

`inbuilt go packages`

`3rd party packages from github`

`project defined packages`

### PR DESCRIPTION
**Pull request by RahmaYasser**

### CODE REVIEW

File: pkg/domain/response.go [hunk 1]
```diff
@@ -1,31 +1,36 @@
...
 	var res response
 	res = response{Message: msg, Status: "500"}
 	return res
 }
 
+func Response401(msg string) response {
+	var res response
+	res = response{Message: msg, Status: "401"}
+	return res
+}
```
Comment (odetolakehinde) [inline pkg/domain/response.go]: The http package has all of these statusCodes predefined. And the Gin package can handle some of these responses too.

File: pkg/domain/user/userRepositoryDb.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+func (r UserRepositoryDb) Login(userObj User) (string,error){
+	row := r.db.QueryRow("select password from users where email=$1", userObj.Email)
+	err := row.Scan(&userObj.Password)
+	if err != nil {
+		return "",err
+	}
+	return userObj.Password,nil
+}
```
Comment (odetolakehinde) [inline pkg/domain/user/userRepositoryDb.go]: how does this login work? uses just the email to return the password or what exactly?

File: pkg/domain/user/userRepositoryDb.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+func (r UserRepositoryDb) Signup(user User) (int,error){
+	query := "insert into users (email, password) values($1, $2) RETURNING id;"
+	err := r.db.QueryRow(query, user.Email, user.Password).Scan(&user.Id)
+	if err != nil {
+		return 0,err
+	}
+	return user.Id,nil
+}
```
Comment (odetolakehinde) [inline pkg/domain/user/userRepositoryDb.go]: The password in my opinion should be hashed first before storage.

File: pkg/domain/user/userRepositoryDb.go (hunk 2 above)
Comment (RahmaYasser) [reply pkg/domain/user/userRepositoryDb.go]: yes

File: pkg/domain/user/userRepositoryDb.go (hunk 3 above)
Comment (RahmaYasser) [reply pkg/domain/user/userRepositoryDb.go]: yes it's hashed in userHandler
here: https://github.com/RahmaYasser/backend/blob/jwt/pkg/handlers/userHandlers.go
at line 35

File: pkg/handlers/userHandlers.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	w.WriteHeader(http.StatusOK)
+	jwtObj:=JWT{Token: token}
+	json.NewEncoder(w).Encode(jwtObj)
+}
+func valid(email string) bool {
+	_, err := mail.ParseAddress(email)
+	return err == nil
+}
+func validateEmailAndPassword(userObj user.User)(int,string){
```
Comment (odetolakehinde) [inline pkg/handlers/userHandlers.go]: func validateEmailAndPassword(userObj user.User)(int, err)

File: .idea/codeStyles/codeStyleConfig.xml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<component name="ProjectCodeStyleConfiguration">
+  <state>
+    <option name="PREFERRED_PROJECT_CODE_STYLE" value="Default" />
+  </state>
+</component>
```
Comment (odetolakehinde) [inline .idea/codeStyles/codeStyleConfig.xml:5]: add this to .gitignore

File: pkg/errs/errors.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+var UnexpectedError = errors.New("unexpected error")
+var InvalidToken = errors.New("invalid token")
+
+type Response struct {
+	Message string `json:"message"`
+	Status int `json:"status"`
+}
+func NewResponse(message string, status int) *Response {
+	return &Response{Message: message, Status: status}
+}
```
Comment (odetolakehinde) [inline pkg/errs/errors.go]: I just merged @MarwanGalal746 's PR on this errors package. Try to pull and update accordingly.

**Review by odetolakehinde — COMMENTED** [review]
Check update the errors package so I can merge.

File: pkg/errs/errors.go (hunk 6 above)
Comment (RahmaYasser) [reply pkg/errs/errors.go]: okay

**Pull request merged**